### PR TITLE
allow migrating paused upkeeps

### PR DESCRIPTION
**Pull request by FelixFan1992**

### CODE REVIEW

**Review by vla33 — DISMISSED** [review]

**Review by infiloop2 — CHANGES_REQUESTED** [review]
Ah i had forgotten about migration. We'll also need to support migrating from 1.2 registry to registry with pause functionality. 

We'll need to update the transcoder version and UpkeepTranscoder.sol to update the field when no paused field is present

**Comment by RyanRHall** [comment]
I think this looks good - we can probably punt on the transcoder work until later

File: contracts/src/v0.8/dev/KeeperRegistryBase.sol [hunk 1]
```diff
@@ -28,7 +28,7 @@ abstract contract KeeperRegistryBase is ConfirmedOwner, ExecutionPrevention, Ree
   uint256 internal constant PPB_BASE = 1_000_000_000;
   uint64 internal constant UINT64_MAX = 2**64 - 1;
   uint96 internal constant LINK_TOTAL_SUPPLY = 1e27;
-  UpkeepFormat internal constant UPKEEP_TRANSCODER_VESION_BASE = UpkeepFormat.V1;
+  UpkeepFormat internal constant UPKEEP_TRANSCODER_VERSION_BASE = UpkeepFormat.V1;
```
Comment (FelixFan1992) [inline contracts/src/v0.8/dev/KeeperRegistryBase.sol:31]: i think this should be V2?

Comment (infiloop2) [reply contracts/src/v0.8/dev/KeeperRegistryBase.sol:31]: yes

File: contracts/test/v0.8/dev/KeeperRegistryDev.test.ts [hunk 2]
```diff
@@ -2285,6 +2286,29 @@ describe('KeeperRegistryDev', () => {
...
+        // verify the upkeep is paused
+        expect((await registry.getUpkeep(id)).paused).to.equal(true)
+        // migrate
+        await registry.connect(admin).migrateUpkeeps([id], registry2.address)
+        expect((await registry.getState()).state.numUpkeeps).to.equal(0)
+        expect((await registry2.getState()).state.numUpkeeps).to.equal(1)
+        expect((await registry.getUpkeep(id)).balance).to.equal(0)
+        expect((await registry2.getUpkeep(id)).balance).to.equal(toWei('100'))
+        expect((await registry.getUpkeep(id)).checkData).to.equal('0x')
+        expect((await registry2.getUpkeep(id)).checkData).to.equal(randomBytes)
```
Comment (infiloop2) [inline contracts/test/v0.8/dev/KeeperRegistryDev.test.ts:2304]: why randomBytes? Should be 0x?

**Review by infiloop2 — CHANGES_REQUESTED** [review]
small inline comments

File: contracts/test/v0.8/dev/KeeperRegistryDev.test.ts (hunk 2 above)
Comment (FelixFan1992) [reply contracts/test/v0.8/dev/KeeperRegistryDev.test.ts:2306]: i think this upkeep is created with checkData = randomBytes.
so after migration, the old upkeep's checkData is deleted and the new upkeep has randomBytes as checkData

File: contracts/src/v0.8/dev/UpkeepTranscoderDev.sol [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+/**
+ * @notice Transcoder for converting upkeep data from one keeper
+ * registry version to another
+ */
+contract UpkeepTranscoderDev is UpkeepTranscoderInterface, TypeAndVersionInterface {
+  error InvalidTranscoding();
+
+  /**
+   * @notice versions:
+   * - UpkeepTranscoder 1.0.0: placeholder to allow new formats in the future
```
Comment (infiloop2) [inline contracts/src/v0.8/dev/UpkeepTranscoderDev.sol:17]: nit. add a line here

File: contracts/test/v0.8/dev/KeeperRegistryDev.test.ts (hunk 2 above)
Comment (infiloop2) [reply contracts/test/v0.8/dev/KeeperRegistryDev.test.ts:2306]: oh yeah i misunderstood before

**Review by infiloop2 — APPROVED** [review]
lgtm

File: contracts/src/v0.8/dev/UpkeepTranscoderDev.sol (hunk 3 above)
Comment (FelixFan1992) [reply contracts/src/v0.8/dev/UpkeepTranscoderDev.sol:18]: Done

**Review by infiloop2 — DISMISSED** [review]

File: contracts/src/v0.8/dev/KeeperRegistryDev.sol [hunk 4]
```diff
@@ -480,20 +480,20 @@ contract KeeperRegistryDev is
...
     // Executed through logic contract
     _fallback();
   }
 
   /**
-   * @inheritdoc MigratableKeeperRegistryInterface
+   * @inheritdoc MigratableKeeperRegistryInterfaceDev
    */
-  UpkeepFormat public constant override upkeepTranscoderVersion = UPKEEP_TRANSCODER_VESION_BASE;
+  UpkeepFormatDev public constant override upkeepTranscoderVersion = UPKEEP_TRANSCODER_VERSION_BASE;
```
Comment (RyanRHall) [inline contracts/src/v0.8/dev/KeeperRegistryDev.sol:493]: ah, here it's public

**Review by RyanRHall — CHANGES_REQUESTED** [review]

File: contracts/src/v0.8/dev/UpkeepTranscoderDev.sol [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+ */
+contract UpkeepTranscoderDev is UpkeepTranscoderInterfaceDev, TypeAndVersionInterface {
+  error InvalidTranscoding();
+
+  /**
+   * @notice versions:
+   * - UpkeepTranscoder 1.3.0: placeholder to be replaced with logic to convert Upkeep V1 to V2
+   * - UpkeepTranscoder 1.0.0: placeholder to allow new formats in the future
+   */
+  string public constant override typeAndVersion = "UpkeepTranscoder 1.3.0";
```
[the remark this answers is not in the document]
Comment (FelixFan1992) [reply contracts/src/v0.8/dev/UpkeepTranscoderDev.sol:18]: Done

**Review by infiloop2 — APPROVED** [review]
looks good

**Comment by infiloop2** [comment]
Based on offline conversation dismissing @RyanRHall's review as his comments are addressed in order to merge

**Pull request merged**